README.txt
First name: Zijie
Last name: Deng
ID: 883871989

This project is about adding player character and Monsters to the dungeon, and make them move around to "kill each other".
 
Specification:
	1. Randomly assign locations for PC and Monsters, which will always be placed in room.
	2. PC is represented by '@', Monsters are represented by char from '0' to '9' and 'a' to 'f'.
	3. goMonsters determines whether monsters have Intelligence, Telepathy, Tunneling Ability, and Erratic behavior or not, and call goDijkstra, goTelepathy, and goRandom respectively.
	4. correctly remove character, when new arrival kills the original occupant.
	5. correctly find path for tunnelers, non-tunnelers, tunnler can move through rocks, others cannot.
	6. end game once PC is killed.

Methods about Character:
	Create point:
		1. typedef struct{}Character;
		2. Character* createCharacter(int);
	Assign Random Locations for characters:
		void assignRandomLocation(Character*);
	Remove character, if new arrival kills the original occupant:
		void removeMonster(Character*);

Methods about Moving in dungeon:	
	The method for all, which takes Character*, determine the abilities:
		void goMonster(Character*);
	The method for going from original point to another point:
		void go(Character*);
	The method for moving randomly, including determine tunneler: 
		void goRandom(Character*);
	The method for intelligent monsters to move:
		void goDijkstra(Character*);
	The method for Telepathy monsters to move:
		void goTelepathy(Character*);

main.c contains:
	(1)In a for loop to check for swiches:
		--nommon
		Example: if user enter  "--nommon NUMBER" after "./rlg327":
			change numMonster to that NUMBER.
	(2)In a while loop to move characters and PC, which also calls pathFinding to update current PC location.

Makefile:
	This makefile helps to compile main.c room.c room.h dungeon.c dungeon.h
	pathFinding.c pathFinding.h character.c character.h
	Example:
		Enter "make" to complie.
		Enter "make clean" to remove "./main" and all file ended with ".o"
		Enter "./rlg327" to run the preject.


CHANGELOG.txt
First name: Zijie
Last name: Deng
ID: 883871989

January 23rd 2016:
	create room.h and room.c
		1. Built Room: typedef struct
		2. The method for randomly generating a room: Room* getRandRoom();
			w represents width, at least 3 units and at most 30 units, 
			h represents hight, at least 2 units and at most 15 units. 
	   		x represents x coordinate of room's top right point.
	   		y represnets y coordinate of room's top right point.
		3. The method for judging whether two rooms can coexist.
			 int coexist(Room* room1, Room* room2);
			 Return 1 if coexist and 0 if not.
	
	Started main.c
		1. Declaring a 2D array: char matrix[80][21];
		2. Creating a Room* and using printf to test getRandRoom();
		3. Creating a method to fill up this 2D array with '+',
			checking if the dungeon is created.
			By using: void constructMatrix();
		4. Creating a method to print out a room randomly generated.
			By using: void placeRoom(Room* room);
		5. Creating a method to print out the whole dungeon.
			By using: void printMatrix();

	Started and finish makefile
		for main.c room.c room.h

January 24th 2016:
	Continue to modify room.h and room.c
		1. The method for checking if two rooms can coexist: 
			int coexist(Room* room1, Room* room2);
			Requiring two rooms aren't the same room, 
			and they cannot contact one another.
			There must be at least 1 cell of non-room 
			between any two different rooms. 
		2. The method for finding the centroid point of a room: 
			void getCentroid(int*, int*, Room*);
			Preparing for building path between two rooms.

	Continue to modify main.c
		1. Moving and defining 2D array matrix outside main.
			char matrix[NUM_ROW][NUM_COL];
		2. Creating an array to store all the rooms. 
			Room* arr[ARR_SIZE];
		3. Creating a method for cheking if current room coexist 
			with all the rooms that exist in arr.
			int coexistAll(Room* room, int number);
			Return 1 if coexist and 0 if not.
		4. Writing a while loop to to limit attempts of trying within 2000,
			whose aim is to fill up the arr with all rooms coexist.
		
		
January 25th 2016:
	Continue to modify room.h and room.c
		1. Put method getCentroid(int*, int*, Room*) into another method.
		2. The method for building path between two rooms.
			void connect(Room* room1, Room* room2, char matrix[NUM_ROW][NUM_COL]);
			(1)Find the centroid points from two rooms.(x1, y1)and(x2,y2).
			(2)Calculate the difference between these two points. x and y.
			(3)Creating an array of int to identify which direction to move
			    by declaring an int as length for an array, 
				which is the absolute value of x plus that of y.
				x is represented by 0, y is represented by 1.
			(4)Randomly choose directions.
				by Writing a helper method to reorganize the array.
				static void shuffle(int length, int* arr)

	Cotinue to modify main.c
		1. Connecing all the rooms with the first room, in a for loop.
		2. Write debug.
		3. Change'+' to ' '.	
	
	Continue to write room.h and room.c
		1. Continue working on method: 
			connect(Room* room1, Room* room2, char matrix[NUM_ROW][NUM_COL]);
			Print out the path with '#' according to the directions,
			which is stored in array.
			And no '#' in rooms.
	
	Finish project.

January 29th 2016:
	1. Create dungeon.c and move and change main.c to dungeon.c as a method:
		int dungeon_random();
	2. Create dungeon.h and move all "#include <>" and "#define" into it.
	3. Create main.c that can check swiches and call methods in dungeon.c:
		(1) Check argv equals to "--load", "--save", "--debug", or an number
		input by user.
		(2) According to the condition, call dungeon_load(char**), 
		dungeon_random().

January 31st 2016:
	1. Implement method dungeon_load(char**): 
		(1) Take a filename.
		(2) read the first 6 bytes with value RLG327.
		(3) read the bytes from 6 to 9 with value 0 as file version.
		(4) read the bytes from 10 to 13 as integer size of the file.
		(5) read the bytes from 14 to 1495 as the hardness value, ranging 
			from 0 to 255, for each element in matrix except boarder, 255.
		(6) read the bytes from 1495 to end, every 4 unsigned 8-bit integer
			as x,y, w, h.
		(7) get room numbers by using size minus 1495.
		(8) write a for loop to the x, y, w, h for each room and place room.		(9) according to harness, the bytes from 14 to 1495, to place path.

February 1st 2016:
	1. add hardness for each char in matrix in dungeon_random():
		(1) create a 2D array h_matrix[21][80] to store hardness.
		(2) random number from 1 to 254 for all the char in matrix.
		(3) put 0 into h_matrix as hardness for rooms and pathes.
	2. Create method for --save: dungeon_save();
	2. Create method for --save: dungeon_save();

February 5th 2016:
	1. Create pathFinding.c
		(1) create point: 
			typedef short Point;
			Point makePoint(char x, char y);
		(2) create method to get x coordinate, and y coordinate:
			char getX(Point);
			char getY(Point);
		
February 6th 2016:
	1. Continue to modify pathFinding.c
		(1) compare the distance of two points:
			the method for non-tunneling monsters:
				int32_t compareN(const void *key, const void *with);
			the method for tunnelers:
				int32_t compareT(const void *key, const void *with);
		(2) create method for assigning distance from pc to the rest of points in dungeon:
			the method for non-tunneling monsters:
				void pathFindingN(Point*);
		(3) create method for choosing the neighbor points around the choosing point:
			Point* getNeighbors(Point);
		(4) create method for getting weight according to hardness:
			int getWeight(Point);

February 12th 2016:
	1. Continue to modify pathFinding.c
		(1) finding problem in method pathFindingN(Point*), which is about address.
		(2) change "typedef short point" to "typedef struct{} Point";
		(3) change the retrun type of makePoint to Point*;
		(3) delet getX(); and getY(); since I can directly use point->x and point->y;
		(4) change Point to Point* in getNeighbor(Point); pathFindingN(Point); and getWeight(Point);	

dungeon.c
#include "dungeon.h"

#include "helper.h"

#include "character.h"

int dungeon_load(char* filepath)
{
	
	FILE* f = fopen(filepath, "r");
	
	if(!f)
	{
		fprintf(stderr, "Could not open file %s\n", filepath);
		return 1;
	}
		
	char buffer[7];
	fread(buffer, sizeof(*buffer), 6, f);	
	buffer[6] = '\0';
	int vers;
	int size;
	fread(&vers, sizeof(int), 1, f);
	fread(&size, sizeof(int), 1, f);
	unsigned char hardness[1482];
	fread(hardness, sizeof(*hardness), 1482, f);	
	int j;

	size = helper_reverseBytes(size);
	printf("dungeon_load: size is: %d\n", size);
	
	int room_num = (size - 1496)/4;

	constructMatrix();

	int r, c;
	for(r=1; r<20; r++)
	{
		for(c=1; c<79; c++)
		{
			matrix[r][c] = ' ';
		}
	}

	for(r=0;r<19;r++)
	{
		for(c=0;c<78;c++)
		{
			h_matrix[r+1][c+1] = hardness[r*78 + c];
			
			if(!hardness[r*78 + c])
				matrix[r+1][c+1] = '#';
		}
	}
	
	int i;
	for(i=0; i<room_num; i++)
	{
		Room* room = malloc(sizeof(*room));

		uint8_t u8b[4]; // unsigned 8-bit integer buffer
		fread(u8b, sizeof(*u8b), 4, f);
		printf("x=%d, y=%d, w=%d, h=%d\n", 
		u8b[0], u8b[1], u8b[2], u8b[3]);

		room->x = u8b[0];
		room->y = u8b[1];
		room->w = u8b[2];
		room->h = u8b[3];

		arr[i] = room;

		placeRoom(room);		
	}
	//printMatrix();
	
	fclose(f);
	   	
	return 0;
}

int dungeon_save(char* filepath)
{
	FILE* f = fopen(filepath, "w");
	
	if(!f)
	{
		fprintf(stderr, "Could not open file %s\n", filepath);
		return 1;
	}
		
	fwrite("RLG327", sizeof(char), 6, f);	
	
	int vers = 0;
	
	int size = room_count * 4 + 1496;
	printf("dungeon_save: size is: %d\n", size);
	size = helper_reverseBytes(size);
	
	fwrite(&vers, sizeof(int), 1, f);
	fwrite(&size, sizeof(int), 1, f);
	
	unsigned char hardness[1482];
	
	int r, c;
	for(r=0;r<19;r++)
	{
		for(c=0;c<78;c++)
		{
			hardness[r*78 + c] = h_matrix[r+1][c+1];
		}
	}
	
	fwrite(hardness, sizeof(*hardness), 1482, f);	
	
	
	int i;
	for(i=0; i<room_count; i++)
	{
		uint8_t u8b[4]; // unsigned 8-bit integer buffer

		u8b[0] = arr[i]->x;
		u8b[1] = arr[i]->y;
		u8b[2] = arr[i]->w;
		u8b[3] = arr[i]->h;

		fwrite(u8b, sizeof(*u8b), 4, f);
   }

   fclose(f);
   
   return 0;
}

/* function from version 1.01 */
int dungeon_random()
{
	int count = 0;
	int currentArr = 0;
	while(count < 2000 && currentArr < ARR_SIZE)
	{	
		Room* roomCurrent = getRandRoom();
		if(coexistAll(roomCurrent, currentArr))
		{
			arr[currentArr++] = roomCurrent;
		}
		else
		{
			free(roomCurrent);
		}
		count++;
	}
	
	printf("seed: %u\n", seed);
	
	// printf("w: %d, h: %d, x: %d, y:%d\n", r->w, r->h, r->x, r->y);
	constructMatrix();
	
	int i;
	for(i=0; i<currentArr; i++)
	{
		placeRoom(arr[i]);
		
		if(debug)
			printf("w: %d, h: %d, x: %d, y:%d\n",
			arr[i]->w, arr[i]->h, arr[i]->x, arr[i]->y);
		
		connect(arr[i], arr[0], matrix);
	}
	
	//printMatrix();

	return 0;
}

int coexistAll(Room* room, int number)
{
	int i;
	for(i=0; i<number; i++)
	{
		if(!coexist(room, arr[i]))
			return 0;
	}
	return 1;
}

void constructMatrix()
{
	int i;
	int j;
	for(i=0; i<NUM_ROW; i++)
	{
		for(j=0; j<NUM_COL; j++)
		{   
			matrix[i][j] = ' ';
			h_matrix[i][j] = 1 + rand() % 254;
		}
	}
	int r, c;
	for(c=0; c<80; c++)
	{
		matrix[0][c] = ' ';
		matrix[20][c] = ' ';

		h_matrix[0][c]=255;
		h_matrix[20][c]=255;
	}
	for(r=0; r<21; r++)
	{
		matrix[r][0] = ' ';
		matrix[r][79] = ' ';

		h_matrix[r][0]=255;
		h_matrix[r][79]=255;
	}
}

void placeRoom(Room* room)
{
	int i;
	int j;
	for(i=room->x; i< room->x+room->w; i++)
	{
		for(j=room->y; j< room->y+room->h; j++)
		{
			matrix[j][i] = '.';
			h_matrix[j][i] = 0; 
		}
	}
	room_count++;
}

void printMatrix()
{
	int i;
	int j;
	for(i=0; i<	NUM_ROW; i++)
	{	
		for(j=0; j< NUM_COL; j++)
		{
			if(!c_matrix[i][j])
				printf("%c", matrix[i][j]);
			/*else if(c_matrix[i][j]->isPC)
				printf("@");*/
			else  {
				printf("%c", getChar(c_matrix[i][j]));				
				// printf("\nchar: <%c>\n", getChar(c_matrix[i][j]));
			}
		}
		printf("\n");
	}
}
